### ISSUE TITLE
panic "out of range for slice"

### ISSUE THREAD

**Issue by antiochp** [mimblewimble/grin#499]
No line numbers unfortunately.
This happened just now on the latest build from `milestone/testnet1`.

This was on my seed - there were a lot of connected peers at the time.
Appears to have seen multiple blocks mined simultaneously and was rolling back from one to another.
Processing block `9e87d22e at 18602`
Rewind to forked block `767892e5 at 18601`

```
Dec 15 16:02:54.299 DEBG handle_payload: Block 9e87d22e
Dec 15 16:02:54.299 DEBG Received block 9e87d22e at 18602 from network, going to process.
Dec 15 16:02:54.300 DEBG pipe: process_block 9e87d22e at 18602 with 0 inputs and 1 outputs.
Dec 15 16:02:54.304 DEBG pipe: validate_header cuckoo size 16
Dec 15 16:02:54.304 DEBG Starting new sumtree extension.
Dec 15 16:02:54.321 DEBG validate_block: forked_block: 767892e5 at 18601
Dec 15 16:02:54.322 DEBG Rewind sumtrees to header 767892e5 at 18601
Dec 15 16:02:54.322 DEBG Rewind sumtrees to output pos: 40064, kernel pos: 38660
Dec 15 16:02:54.351 DEBG -- outputs --
thread '<unnamed>' panicked at 'index 2603965 out of range for slice of length 2603900', /checkout/src/libcore/slice/mod.rs:748:4
Dec 15 16:02:54.382 DEBG monitor_peers: 1 most_work_peers, 12 connected, 333 total known
Dec 15 16:02:54.382 DEBG monitor_peers: all 333 = 20 healthy + 2 banned + 311 defunct
Dec 15 16:02:54.382 DEBG cleaning V4(54.152.52.242:13414), not connected
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at /checkout/src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/sys_common/backtrace.rs:60
             at /checkout/src/libstd/panicking.rs:381
   3: std::panicking::default_hook
             at /checkout/src/libstd/panicking.rs:397
   4: std::panicking::rust_panic_with_hook
             at /checkout/src/libstd/panicking.rs:611
			 5: std::panicking::begin_panic
		               at /checkout/src/libstd/panicking.rs:572
		     6: std::panicking::begin_panic_fmt
		               at /checkout/src/libstd/panicking.rs:522
		     7: rust_begin_unwind
		               at /checkout/src/libstd/panicking.rs:498
		     8: core::panicking::panic_fmt
		               at /checkout/src/libcore/panicking.rs:71
		     9: core::slice::slice_index_len_fail
		               at /checkout/src/libcore/slice/mod.rs:748
		    10: grin_store::sumtree::AppendOnlyFile::read
		    11: <grin_store::sumtree::PMMRBackend<T> as grin_core::core::pmmr::Backend<T>>::get
		    12: grin_chain::sumtree::Extension::dump
		    13: grin_chain::sumtree::Extension::rewind
		    14: grin_chain::pipe::process_block
		    15: grin_chain::chain::Chain::process_block
		    16: <grin_grin::adapters::NetToChainAdapter as grin_p2p::types::ChainAdapter>::block_received
		    17: <grin_p2p::peers::Peers as grin_p2p::types::ChainAdapter>::block_received
		    18: <grin_p2p::peer::TrackingAdapter as grin_p2p::types::ChainAdapter>::block_received
		    19: std::panicking::try::do_call
		    20: __rust_maybe_catch_panic
		               at /checkout/src/libpanic_unwind/lib.rs:99
		    21: <futures_cpupool::MySender<F, core::result::Result<<F as futures::future::Future>::Item, <F as futures::future::Future>::Error>>
		   as futures::future::Future>::poll
		    22: futures::task_impl::std::Run::run
```

**Comment by ignopeverell** [comment]
Thanks, I may be able to ferret it out still. Seems that it's logging actually, the sumtree `dump` function prints the latest tree nodes when rewinding.

**Comment by ignopeverell** [comment]
I don't think I can tell what happened without seeing what you had in your `sumtree/utxo` folder. You rewound to position 40064 which would be 2,604,160 bytes of data. But somehow it looks like your file was smaller than that. Only think I can think of is it hasn't been resync'd since the last pmmr bug fix but that's unlikely. So please if it happens again backup your `sumtree` dir before restarting.

**Comment by ignopeverell** [comment]
Found out it happened on my seed as well. Going to do some more analysis. Stack with line numbers:

```
thread '<unnamed>' panicked at 'index 2603965 out of range for slice of length 2603900', /checkout/src/libcore/slice/mod.rs:748:4
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:                                                                        
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
             at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::_print
             at /checkout/src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at /checkout/src/libstd/sys_common/backtrace.rs:60
             at /checkout/src/libstd/panicking.rs:381
   3: std::panicking::default_hook                                               
             at /checkout/src/libstd/panicking.rs:397
   4: std::panicking::rust_panic_with_hook    
             at /checkout/src/libstd/panicking.rs:611                                                     
   5: std::panicking::begin_panic                                         
             at /checkout/src/libstd/panicking.rs:572                                                     
   6: std::panicking::begin_panic_fmt                                                                        
             at /checkout/src/libstd/panicking.rs:522                                                             
   7: rust_begin_unwind                                                                            
             at /checkout/src/libstd/panicking.rs:498                                                             
   8: core::panicking::panic_fmt  
             at /checkout/src/libcore/panicking.rs:71
   9: core::slice::slice_index_len_fail
             at /checkout/src/libcore/slice/mod.rs:748 
  10: <core::ops::range::Range<usize> as core::slice::SliceIndex<[T]>>::index
             at /checkout/src/libcore/slice/mod.rs:879
  11: core::slice::<impl core::ops::index::Index<I> for [T]>::index
             at /checkout/src/libcore/slice/mod.rs:730
  12: grin_store::sumtree::AppendOnlyFile::read                                       
             at store/src/sumtree.rs:96                                                                           
  13: <grin_store::sumtree::PMMRBackend<T> as grin_core::core::pmmr::Backend<T>>::get                                                                                   
             at /root/grin/store/src/sumtree.rs:338                                                       
  14: <grin_core::core::pmmr::PMMR<'a, T, B>>::get           
             at /root/grin/core/src/core/pmmr.rs:353                                                      
  15: <grin_core::core::pmmr::PMMR<'a, T, B>>::dump                                         
             at /root/grin/core/src/core/pmmr.rs:410                                                             
  16: grin_chain::sumtree::Extension::dump              
             at chain/src/sumtree.rs:401                                                                     
  17: grin_chain::sumtree::Extension::rewind
             at chain/src/sumtree.rs:371                                                                        
  18: grin_chain::pipe::validate_block     
             at chain/src/pipe.rs:268                                                                        
  19: grin_chain::pipe::process_block::{{closure}}                             
             at chain/src/pipe.rs:93                                                                             
  20: grin_chain::sumtree::extending   
             at chain/src/sumtree.rs:160                                                                         
  21: grin_chain::pipe::process_block
             at chain/src/pipe.rs:92                                                                      
  22: grin_chain::chain::Chain::process_block      
             at chain/src/chain.rs:149                                                                    
  23: <grin_grin::adapters::NetToChainAdapter as grin_p2p::types::ChainAdapter>::block_received
             at grin/src/adapters.rs:77
  24: <grin_p2p::peers::Peers as grin_p2p::types::ChainAdapter>::block_received
             at p2p/src/peers.rs:318
  25: <grin_p2p::peer::TrackingAdapter as grin_p2p::types::ChainAdapter>::block_received
             at p2p/src/peer.rs:241
```

**Comment by ignopeverell** [comment]
So here is what happened:

1. Fork at 18597
2. Main chain continues normally from 18598 to 18601.
3. New block at 18598 continuing the fork accumulates more work than at 18601, rewind occurs and we have a new head at 18598.
4. Another block get mined at 18601, forking from 18600 against the now old fork.

At step 3, the sumtree storage got truncated back to 18598. On step 4, we detect properly that it's a fork but as the ancestry is based on positions, and there still is a 18600 in store despite the rewind (that's the bug), we decide that we only need to rewind 1 position and replay from 18600. Which doesn't exist because we went back to 18598 in the meantime.

This likely wouldn't arise if #280 was fixed, as a couple blocks couldn't accumulate more work than 4 other blocks.
